lsコマンドを作るオブジェクト指向版

rira100000000 opened the pull request:
[lsコマンドを作るオブジェクト指向版](https://bootcamp.fjord.jp/practices/213)の提出物です。

rira100000000 committed:
lsコマンドにクラスを導入

lsコマンドにクラスを導入しました。

JunichiIto reviewed (commented):
動画でレビューしたので別途リンクを送ります〜。

rira100000000 committed:
empty?メソッドを使用するように修正

配列の空判定でempty?メソッドを使用するように修正しました。

rira100000000 committed:
RuboCop対応

.rubocop.ymlにTargetRubyVersionを設定しました。

rira100000000 committed:
クラスの再実装

listクラスを削除し、separetorクラスを実装しました。
オブジェクト指向を意識した実装に修正しました。

rira100000000 committed:
empty?メソッドを使用するように修正

array == [] の条件式を array.empty?に修正しました

rira100000000 committed:
文字列連結にjoinを使用するように修正

連結したい文字列を配列にしてjoinするようにしました

rira100000000 committed:
{}が複数行の時の体裁修正

{}が複数行になる時の{}の位置を修正しました。

rira100000000 committed:
コメントの追加

JunichiIto commented on 07.ls_object/ls.rb:
```diff
@@ -0,0 +1,33 @@
+def main
+  paths, options = parse_option
+  if options[:l]
+    if paths.empty?
+      puts LongFormatList.new.make_long_format_list(options)
+    else
+      puts LongFormatList.new.make_long_format_list_with_paths(paths, options)
+    end
```
```suggestion
    puts LongFormatList.new.hogehoge(paths, options)
```

のようにしてオブジェクト自身が賢く状況を判断できる方がオブジェクト指向的にはベターです。
（クラスを利用する人が気を遣わなくて済むので楽にプログラムが書けるようになる）

rira100000000 replied:
`make_long_format_list(paths, options)`の中で処理を分岐させます。

JunichiIto commented on 07.ls_object/short_format_list.rb:
```diff
@@ -0,0 +1,73 @@
+    end
+  end
+
+  def list_directory_files_for_display(directory_list, flag, need_reverse_order)
```
privateメソッドとpublicメソッドを明確に分けましょう。
そうすることでプログラム修正時に影響の範囲が特定しやすくなります。

rira100000000 replied:
privateメソッドとpublicで分割します

JunichiIto commented on 07.ls_object/short_format_list.rb:
```diff
@@ -0,0 +1,73 @@
+      display_lines = adjust_list_for_display(file_list)
+      display_lines.each { |line| result << line }
+      result << "\n" unless file_list.empty?
+      directorys = list_directory_files_for_display(separator.fetch_directory(options[:r]), flag, options[:r])
```
複数形のスペルが違いますね。
https://www.wordhippo.com/what-is/the-plural-of/directory.html

rira100000000 replied:
linkありがとうございます。
修正します。

JunichiIto commented on 07.ls_object/short_format_list.rb:
```diff
@@ -0,0 +1,73 @@
+    lines = []
+    max_file_names = []
+    files.each_with_index do |file_name, i|
+      now_row = i % rows
```
「現在の」という意味ではcurrentがよく使われます

rira100000000 replied:
currentに修正します。

JunichiIto commented on 07.ls_object/short_format_list.rb:
```diff
@@ -0,0 +1,73 @@
+    lines.map do |file_names|
+      display_line = []
+      file_names.each_with_index do |file_name, i|
+        display_line << "#{file_name}#{' ' * (max_file_names[i] - calc_file_name_size(file_name) + SPACE_FOR_COLUMNS)}"
```
この行、ちょっと読みにくいですね。。
https://twitter.com/jnchito/status/1326468865103073280

rira100000000 replied:
ごちゃごちゃしてますね…
スペースの計算と`ljust`で処理を分けるよう修正します。

JunichiIto commented on 07.ls_object/short_format_list.rb:
```diff
@@ -0,0 +1,73 @@
+COLUMNS = 3
+SPACE_FOR_COLUMNS = 2
+
+class ShortFormatList
```
（実装コード全体を見て）
条件分岐が多数あるわりには、short formatのスクショが1種類しかなく、テストが不足してるな〜という印象を受けました。

参考 [書いたらテストしろ、テストしないなら書くな！ \| FBC](https://bootcamp.fjord.jp/pages/write-code-write-test)

「複雑な仕様を盛り込むならそれを全部テストする」or「テストするのが面倒なら必須要件を最低限満たすシンプルな仕様で実装する」のどちらかでお願いします。
今回はオブジェクト指向プログラミングの課題であってlsコマンドの作り込みの課題ではないので、個人的には後者の方がいいかなーと思います（その方がレビューするのもラクですし・・・）。

rira100000000 replied:
lsコマンドの課題じゃないので、パスの指定は一つまでにします。

JunichiIto commented on 07.ls_object/short_format_list.rb:
```diff
@@ -0,0 +1,73 @@
+      separator = Separator.new(paths)
+      file_list = separator.fetch_file(options[:r])
+      display_lines = adjust_list_for_display(file_list)
+      display_lines.each { |line| result << line }
```
`display_lines`は不要な変数なのでは？

rira100000000 replied:
プログラム全体で不要な変数の見直しをします。

JunichiIto commented on 07.ls_object/separator.rb:
```diff
@@ -0,0 +1,17 @@
+# frozen_string_literal: true
+
+class Separator
```
クラス名から想像した実装と、実際の実装が大きく異なっていたので、命名を見直した方がいいかも、と思いました。

rira100000000 replied:
`Separator`クラスはpathの複数指定をサポートしなくなった結果不要になったため削除します。

命名は`FileDirectoryDivider`など考えたのですがどうでしょうか？

JunichiIto replied:
定義されたメソッドがfetch_fileやfetch_directoryなので、DataFetcherとかですかね〜。
より具体的にするならFileOrDirectoryFetcherとかになりそうですが、いまいちイケてないですよね。
パシッといい名前が付けられない場合はそもそもクラスやモジュールの設計が悪い、ということもよくあります。

rira100000000 replied:
なるほどです！
名前がちゃんと付けられるかどうかは設計の重要な指標になりそうです。

JunichiIto commented on 07.ls_object/separator.rb:
```diff
@@ -0,0 +1,17 @@
+
+  def fetch_file(need_reverse_order = nil)
+    result = @paths.select { |path| File::Stat.new(path).file? }.sort
+    need_reverse_order ? result.reverse! : result
```
!無しの方がいいですね。

[\!が付く破壊的メソッドの使用は原則として避ける \| FBC](https://bootcamp.fjord.jp/pages/do-not-use-destructive-method)

rira100000000 replied:
`reverse`を使うよう修正します

JunichiIto commented on 07.ls_object/separator.rb:
```diff
@@ -0,0 +1,17 @@
+    @paths = paths
+  end
+
+  def fetch_file(need_reverse_order = nil)
```
true/falseを想定しているなら、

```suggestion
  def fetch_file(need_reverse_order = false)
```

の方がわかりやすいと思います。

あと「〜する必要がある」はrequiredもよく使われます。
`reverse_required`ぐらいが短くて良いかもしれません。

rira100000000 replied:
`false`を使うようにします。

`need_reverse_order`は`reverse_required`に変更します。

JunichiIto commented on 07.ls_object/long_format_list.rb:
```diff
@@ -0,0 +1,100 @@
+      file_detail
+    end
+    long_format_file_list = list_long_format_for_display(file_details)
+    long_format_file_list.each { |file| result << file }
```
`long_format_file_list`は不要な変数です。
プログラム全体を通して、他にも不要な変数がないかよく見直してください。
（「いったいここで何をやってるんだろう？？この処理はここで必要なんだろうか？」と30行目以降のコードを何度も見直しましたが、結局使われてなかったので「えええーっ」となりました。このように不要なコードがあると、コードの読み手に無駄な時間を使わせることになります）

rira100000000 replied:
お手間をかけて申し訳ないです！プログラム全体で変数を見直します。
コードが横に長くなるのを嫌ってこういった変数を作ってしまっていましたが、読み手のストレスになるという観点がありませんでした。以後気を付けたいと思います。

↓の`flag`は再利用しませんが、ないと見づらい気がするのですがどうでしょうか？

### 変数flagあり
```ruby
flag = options[:a] ? File::FNM_DOTMATCH : 0
file_name_list = Dir.glob('*', base: Dir.pwd, flags: flag).sort
```

### 変数flagなし
```ruby
file_name_list = Dir.glob('*', base: Dir.pwd, flags: options[:a] ? File::FNM_DOTMATCH : 0).sort
```

JunichiIto replied:
> ↓のflagは再利用しませんが、ないと見づらい気がするのですがどうでしょうか？

そうですね、flagありの方が読みやすいです。

JunichiIto commented on 07.ls_object/long_format_list.rb:
```diff
@@ -0,0 +1,100 @@
+    flag = need_hidden_file ? File::FNM_DOTMATCH : 0
+    result = []
+    directory_list.each do |directory|
+      file_list = Dir.glob('*', base: directory, flags: flag).sort
```
このへんの処理は`make_long_format_list`とほぼ同じになるのではないでしょうか？

そもそもの話ですが、ディレクトリを複数指定できるケースを本当にサポートするのかどうか、という点から検討するのが良いかなと思います。

[書いたらテストしろ、テストしないなら書くな！ \| FBC](https://bootcamp.fjord.jp/pages/write-code-write-test)

rira100000000 replied:
今回はlsコマンド自体のプラクティスではないので、パスの指定は１つまでにします。

パス指定を1つにすると全体的に修正が必要なので、共通化できそうな処理を見直します。

JunichiIto commented on 07.ls_object/long_format_list.rb:
```diff
@@ -0,0 +1,100 @@
+  end
+
+  def compare_max_length(file)
+    @max_length_hash[:link_max_char_length] = [file.detail[:nlink].to_s.length, @max_length_hash[:link_max_char_length]].max
```
```ruby
formatter = LongFormatList.new
formatter.make_long_format_list_with_paths(paths, options)
formatter.make_long_format_list_with_paths(other_paths, options)
```

のような呼ばれ方をすると、pathsで更新された`@max_length_hash`がother_pathsにも引き継がれてしまいそうです。

参考 [（あなたの周りでも見かけるかもしれない）インスタンス変数の間違った使い方 \- Qiita](https://qiita.com/jnchito/items/49fe82364bca2f4f37f5)

rira100000000 replied:
すいません、ここの設計が変わり、`file_detail`を持つ`file_list`クラスを作り、`max_length_hash`を`file_list`クラスのインスタンス変数にしました。
`file_list`が`new`されるたびに新しい`max_length_hash`が作られるので引き継がれる問題はなくなったと思います。

設計を変えずに問題を解決する場合、`LongFormatList`に`calc_max_length`メソッドを作り`make_long_format_list_with_paths`の中で毎回新しくローカル変数として`max_length_hash`を作る方法を考えたのですが、これで解決できるでしょうか？

JunichiIto replied:
> 設計を変えずに問題を解決する場合...これで解決できるでしょうか？

実際の実装を見ないと確定的なことは言えませんが、なんとなくそれでいいような気がします！

JunichiIto commented on 07.ls_object/long_format_list.rb:
```diff
@@ -0,0 +1,100 @@
+  def list_long_format_for_display(file_details)
+    file_details.map do |file|
+      "#{file.detail[:type]}#{file.detail[:mode]} "\
+      "#{file.detail[:nlink].to_s.rjust(@max_length_hash[:link_max_char_length])} "\
```
キーが長すぎるので、これぐらいで十分な気がします。

```suggestion
      "#{file.detail[:nlink].to_s.rjust(@max_length_hash[:nlink])} "\
```

rira100000000 replied:
簡潔な名前に修正します。

JunichiIto commented on 07.ls_object/long_format_list.rb:
```diff
@@ -0,0 +1,100 @@
+      "#{file.detail[:gid].rjust(@max_length_hash[:group_name_max_char_length])} "\
+      "#{file.detail[:size].to_s.rjust(@max_length_hash[:file_size_max_char_length])} "\
+      "#{file.detail[:mtime].strftime('%b %e %R')} "\
+      "#{file.name}\n"
```
文字列の構築ですが、いったん配列に入れてjoinする、という方式の方が読みやすいかもしれません。

```ruby
cols = []
cols << "foo"
cols << "bar"
cols << "baz"
cols.join(" ")
```

joinの引数を"\n"にすれば、行方向への連結も実現できます。

rira100000000 replied:
`join`と配列を使った方法に修正します。

JunichiIto commented on 07.ls_object/long_format_list.rb:
```diff
@@ -0,0 +1,100 @@
+  end
+
+  def make_long_format_list_with_paths(paths, options)
+    result = +''
```
`frozen_string_literal: true`を付けるのなら、文字列は原則イミュータブルにしたいですね。
下のコメントに書いた「行方向への連結」のテクニックを使ってみましょう。

rira100000000 replied:
文字列はイミュータブルで使用するようにします。

JunichiIto commented on 07.ls_object/long_format_list.rb:
```diff
@@ -0,0 +1,100 @@
+    end
+    result = list_long_format_for_display(file_details)
+
+    result.unshift("total #{calc_total_block(file_details)}")
```
unshiftは直感的ではないので、配列なら素直に上から順番にpushしていくロジックに修正しましょう。

rira100000000 replied:
`push`か`<<`を使うようにします。

JunichiIto commented on 07.ls_object/long_format_list.rb:
```diff
@@ -0,0 +1,100 @@
+    result[..-2]
+  end
+
+  def reset_max_length_hash
```
resetでメソッド名が始まると、「何かを **リセットする** ためのメソッド」のように見えます。
ですが、実態は「ハッシュを **返す** メソッド」なので、命名としてちぐはぐな印象を受けます。
そもそもの話として、このメソッドはinitializeの中でしか使われていないので、メソッド化せずにベタ書きするのも一手です。

rira100000000 replied:
別の名前なら`initialize_max_length_hash`とかでしょうか？

今回はべた書きに修正したいと思います。

JunichiIto commented on 07.ls_object/long_format_list.rb:
```diff
@@ -0,0 +1,100 @@
+
+  def list_long_format_for_display(file_details)
+    file_details.map do |file|
+      "#{file.detail[:type]}#{file.detail[:mode]} "\
```
ハッシュだと

```ruby
"#{file.detail[:typo]}#{file.detail[:mode]} "\
```

のようにtypoしてもエラーにならず、nilが返るだけなので、不具合に気付きにくいです。

```ruby
"#{file.type}#{file.mode} "\
```

のようにメソッド化すれば、typoにも気付きやすくなります。
（チェリー本7.2.1項参照）

rira100000000 replied:
インスタンス変数化して、`attr_reader`でメソッド化します。

これは同じくハッシュを使っている`max_length_hash`の各値もメソッド化するべきでしょうか？
ハッシュを使う場合とメソッド化する場合の判断基準はありますか？

JunichiIto replied:
> これは同じくハッシュを使っているmax_length_hashの各値もメソッド化するべきでしょうか？
ハッシュを使う場合とメソッド化する場合の判断基準はありますか？

判断基準はオブジェクトとして扱いたくなるかどうか、かな〜と思います。
たとえば、

```ruby
alice = {
  name: 'Alice',
  age: 20,
  height: 160
}
bob = {
  name: 'Bob',
  age: 24,
  height: 170
}
carol = {
  # ...略
}
```

みたいなハッシュデータがあれば、「これってハッシュとして保持するのではなく、共通の型（入れ物）を用意して、Personという名前で管理した方が堅牢でわかりやすくなるのでは！？」という発想が出てくるかもしれません。

```ruby
# ハッシュよりオブジェクトとして扱う方がなんか良さそう！
alice = Person.new(name: 'Alice', age: 20, height: 160)
bob = Person.new(name: 'Bob', age: 24, height: 170)
carol = Person.new(...)
```

`max_length_hash`についてはどうか、ご自身で考えてみてください〜。

rira100000000 replied:
`max_length_hash`について私の考えをまとめます。
## クラスについて
* クラスを作るときはデータのまとまりで考える。
* 堅牢性を高めるという観点ではメソッド化したい。
* 『データに対する振る舞い、メソッドを定義したかったらクラスを作る』の考え方だと今回は必要なさそう。

## ハッシュについて
* 現時点の実装だとlong_format_listを複数用意することはないのでハッシュでいい気がする。
* ハッシュだと`hash[:something]`で呼び出すのでタイポなどの影響を受ける
 
## Structについて
* `struct.something`のように呼び出せる。
* 振る舞いを持たせたいわけじゃないけどメソッド的に呼び出せるので今回の実装にちょうどいい

以上のように考えたため`Struct`として定義するよう修正します。

JunichiIto reviewed (commented):
コメントしました〜。

rira100000000 committed:
long_formatの分岐処理をクラス内で行うように修正

make_long_format_listとmake_long_format_list_with_pathsの分岐を呼び出し側が意識しなくていいように修正しました。

rira100000000 committed:
不要な変数の削除

再利用しない変数を削除しました。

rira100000000 committed:
変数名とキー名の修正

変数名とキー名を正しいスペル、簡潔な名前に修正しました。

rira100000000 committed:
変数名の修正

rira100000000 committed:
条件式でnilではなくfalseを使うよう修正

true,falseを想定しているため、nilではなくfalseを使うように修正しました。

rira100000000 committed:
reverse!をrecerseに修正

破壊的メソッドを使用しないように修正しました。

rira100000000 committed:
display_lineへの代入処理のリファクタリング

ljstを使うように修正しました。

rira100000000 committed:
修正漏れの修正

キーワード引数にした際に呼び出し側を修正していなかったため、修正しました。

rira100000000 committed:
文字列の構築を配列とjoinで行うように修正

文字列の構築が読みづらかったため、配列とjoinを使うように修正しました。

rira100000000 committed:
仕様、設計の変更

パスの指定は1つまでにしました。
separatorクラスを削除しました。
FileListクラスを導入し、設計が以下のようになりました
* FileListクラスはファイルの詳細であるFileDetailを持っている
* LongFormatListクラス、ShortFormatListクラスはそれぞれFileListクラスを使ってフォーマットに沿ったリストを作る

rira100000000 committed:
reset_max_length_hashメソッドを削除

再利用しないのでinitializeに直接書くようにしました。

rira100000000 committed:
LongFormatListとShortFormatListの共通処理をモジュール化

共通の処理をFetchFileDetailモジュールにmake_formatted_listメソッドとして実装しました

rira100000000 committed:
FileDetailの各値をメソッド化

FileDetailの持つ各値をインスタンス変数化、メソッドで呼び出し可能にしました。

rira100000000 committed:
privateとpublicを分ける

各クラスのメソッドをprivateとpublicに分けました。

rira100000000 committed:
RuboCop対応

rira100000000 committed:
不要なインスタンス変数をローカル変数化

@pathは外部から呼び出さないのでローカル変数にしました。

JunichiIto commented on .rubocop.yml:
```diff
@@ -1,3 +1,6 @@
     - "config/rubocop.yml"
+
+AllCops:
+  TargetRubyVersion: 3.1.1
```
質問）これってバージョン指定しないと何か不都合がありました？

rira100000000 replied:
以前エラーメッセージが出ていたのですが、現在は出ていなかったので削除します。

JunichiIto commented on 07.ls_object/fetch_file_detail.rb:
```diff
@@ -0,0 +1,15 @@
+# frozen_string_literal: true
+
+module FetchFileDetail
+  def make_formatted_list(paths, options)
```
format自体が「整形する」という動詞の意味があるので、`format_list`でもいいかもしれません。

rira100000000 replied:
`format_list`に修正します。

JunichiIto commented on 07.ls_object/short_format_list.rb:
```diff
@@ -0,0 +1,51 @@
+require_relative 'fetch_file_detail'
+
+COLUMNS = 3
+SPACE_FOR_COLUMNS = 2
```
グローバルな定数宣言は避けましょう（と以前伝えた記憶が）

rira100000000 replied:
すいません、以前も指摘されてます🙇‍♂️
クラス内で宣言するようにします。

JunichiIto commented on 07.ls_object/short_format_list.rb:
```diff
@@ -0,0 +1,51 @@
+
+  def fetch_file_details(path, reverse_required, hidden_file_required)
+    file_list = FileList.new(path, reverse_required, hidden_file_required)
+    adjust_list_for_display(file_list.name_list)
```
メソッド名は「fetchする＝取ってくるだけ」なのに、表示用の加工もやろうとしているのでメソッドの責務が複数になり、メソッド名と実装が乖離していますね。

以下の記事を参考にしたりしながら修正してみてください。

[マトリョーシカ人形のようなメソッド設計を避ける \- give IT a try](https://blog.jnito.com/entry/2022/11/28/072945)

rira100000000 replied:
`fetch_file_details`では取得処理のみ行うようにします。

JunichiIto commented on 07.ls_object/short_format_list.rb:
```diff
@@ -0,0 +1,51 @@
+      max_file_names[current_column] ||= 0
+
+      lines[current_row] << file_name
+      max_file_names[current_column] = calc_file_name_size(file_name) if max_file_names[current_column] < file_name.size
```
```suggestion
      max_file_names[current_column] = calc_file_name_size(file_name) if max_file_names[current_column] < calc_file_name_size(file_name.size)
```

にしないと、最大値の判定を間違ってしまうのでは？と思いました。

rira100000000 replied:
判定間違えてました。
`calc_file_name_size(file_name)`に修正します。

JunichiIto commented on 07.ls_object/short_format_list.rb:
```diff
@@ -0,0 +1,51 @@
+
+  def calc_file_name_size(file_name)
+    file_name.each_char.sum do |char|
+      char.ascii_only? ? 1 : 2
```
提出物のスクショを見ると、全部アスキー文字のテストパターンしかないように思いました。

JunichiIto commented on 07.ls_object/short_format_list.rb:
```diff
@@ -0,0 +1,51 @@
+    files.each_with_index do |file_name, i|
+      current_row = i % rows
+      current_column = i / rows
+      lines[current_row] = [] if current_column.zero?
```
このあたりのロジックはtransposeメソッドを活用するともう少しシンプルになりそうですが、検討されましたか？

https://docs.ruby-lang.org/ja/latest/method/Array/i/transpose.html

JunichiIto commented on 07.ls_object/short_format_list.rb:
```diff
@@ -0,0 +1,51 @@
+      current_row = i % rows
+      current_column = i / rows
+      lines[current_row] = [] if current_column.zero?
+      max_file_names[current_column] ||= 0
```
```ruby
max_file_names = Array.new(0)
```

のように初期化するとこの行は不要になるかも？

JunichiIto commented on 07.ls_object/short_format_list.rb:
```diff
@@ -0,0 +1,51 @@
+      display_line = []
+      file_names.each_with_index do |file_name, i|
+        space_count = max_file_names[i] + SPACE_FOR_COLUMNS
+        display_line << file_name.ljust(space_count)
```
mapを活用するといいかも？と思いました。（チェリー本の配列の章を復習してみてください）

JunichiIto commented on 07.ls_object/file_list.rb:
```diff
@@ -0,0 +1,50 @@
+            else
+              directory = ''
+              [@path]
+            end
```
このif文ではfilesとdirectoryという二つのローカル変数に代入しているのに、それぞれ別の方法で代入されているのが不自然に感じました。

JunichiIto commented on 07.ls_object/file_list.rb:
```diff
@@ -0,0 +1,50 @@
+  end
+
+  def list_detail
+    @name_list.map do |file|
```
name list = nameの配列 = 配列は複数形でいいので `@names` になるのでは？と思いました。
そうなるとループ処理の名前がちぐはぐでちょっと違和感ありです。

https://bootcamp.fjord.jp/pages/readable-variable-name#-2

JunichiIto commented on 07.ls_object/file_list.rb:
```diff
@@ -0,0 +1,50 @@
+    [file_list, directory]
+  end
+
+  def compare_max_length(file)
```
compareは「AとBを比較する（で、その結果がどうだったかを戻り値で返す）」という動作が期待されるので、メソッド名としてちょっと違和感ありです。

rira100000000 replied:
`update_max_length`に修正します。

JunichiIto commented on 07.ls_object/file_list.rb:
```diff
@@ -0,0 +1,50 @@
+
+  def compare_max_length(file)
+    @max_length_hash[:nlink] = [file.nlink.to_s.length, @max_length_hash[:nlink]].max
+    @max_length_hash[:uid] = [Etc.getpwuid(file.uid.to_i).name.to_s.length, @max_length_hash[:uid]].max
```
FileDetailクラスでは

```ruby
@uid = Etc.getpwuid(stat.uid).name
```

となっていますが、そうすると、ここの `Etc.getpwuid(file.uid.to_i).name` で返ってくる値は何？と思いました。

JunichiIto commented on 07.ls_object/short_format_list.rb:
```diff
@@ -0,0 +1,51 @@
+    adjust_list_for_display(file_list.name_list)
+  end
+
+  def adjust_list_for_display(files)
```
```ruby
# 呼び出す側
adjust_list_for_display(file_list.name_list)

# メソッド側
def adjust_list_for_display(files)
```

となっていて引数の名前が揃ってないのが気になりました。

https://bootcamp.fjord.jp/pages/readable-variable-name#-4

JunichiIto commented on 07.ls_object/fetch_file_detail.rb:
```diff
@@ -0,0 +1,15 @@
+# frozen_string_literal: true
+
+module FetchFileDetail
```
モジュール名は動詞で始まるより、動詞+able で終わる方が多いですね。

rira100000000 replied:
`FileDetailFetchable`に修正します。

JunichiIto commented on 07.ls_object/long_format_list.rb:
```diff
@@ -0,0 +1,43 @@
+class LongFormatList
+  include FetchFileDetail
+
+  private
```
publicメソッドがひとつもなく、いきなりprivateで始まるのは不自然ですね。
実際はFetchFileDetailにpublicメソッドが定義されているわけですが、それであれば

```ruby
class LongFormatList < SomeClass
```

のように継承を使っている方がオブジェクト指向的にまだ自然です。

ただ、個人的な考えではモジュールをincludeしてLongFormatListを「ファイルも取得するし、フォーマットもするクラス」としてしまうとクラスの責務が複数になるので、それよりも「フォーマットするだけのクラス（＝フォーマットの専門家）」とした方がいいと思います（クラスの責務はなるべく一つにする）。

ファイルの取得はクラスの外部で行って、LongFormatListは「このファイルを渡すからあとはフォーマットよろしく！」とお願いできるようなクラスになっているイメージです。（ShortFormatListも同様）

rira100000000 replied:
`FileDetailfetcher`クラスを作り、ファイル情報の取得を行います。
`LongFormatList`、`ShortFormatList`はフォーマットだけ行うようにします。

JunichiIto commented on 07.ls_object/long_format_list.rb:
```diff
@@ -0,0 +1,43 @@
+  private
+
+  def fetch_file_details(path, reverse_required, hidden_file_required)
+    result = []
```
変数の寿命が長いですね。

[変数の寿命はできるだけ短くしよう \| FBC](https://bootcamp.fjord.jp/pages/311#comment_115576)

JunichiIto commented on 07.ls_object/long_format_list.rb:
```diff
@@ -0,0 +1,43 @@
+    result = []
+    file_list = FileList.new(path, reverse_required, hidden_file_required)
+
+    file_details = file_list.list_detail
```
左辺と右辺で名前が違うのが違和感有りです。
同じデータには同じ名前を付けないと読み手が混乱する原因になります。

[ユーザーを見つけたら……Aさん「自転車に入れます」Bさん「人間に入れます」Cさん「ユーザーに入れます」 \| FBC](https://bootcamp.fjord.jp/pages/readable-variable-name)

JunichiIto commented on 07.ls_object/long_format_list.rb:
```diff
@@ -0,0 +1,43 @@
+
+    file_details = file_list.list_detail
+    result << "total #{calc_total_block(file_details)}"
+    result << adjust_list_for_display(file_details, file_list.max_length_hash)
```
max_length_hashはこのクラスでしか使わないので、LongFormatListが自前で持つべきデータかなと思います。

rira100000000 replied:
`LongFormatList`で持つように修正します。

JunichiIto commented on 07.ls_object/file_detail.rb:
```diff
@@ -0,0 +1,58 @@
+    @gid = Etc.getgrgid(stat.gid).name
+    @size = stat.size
+    @mtime = stat.mtime
+    @blocks = stat.blocks
```
インスタンス変数はスコープの広い変数なので、無駄に増やさない方がいいです。

たとえば、

```ruby
class CreditCard
  attr_reader :name, :age 

  def initialize(user)
    @name = user.name
    @age = user.age
  end
end
```

というクラスがあった場合、nameもageもuserのデータさえあれば導出可能なので、以下のように修正する方がベターです。
（nameやageメソッドが遅いメソッドの場合は、パフォーマンスを優先して＝データをキャッシュする目的で上のように書く場合もあります）

```ruby
class CreditCard
  def initialize(user)
    @user = user
  end

  def name
    @user.name
  end

  def age
    @user.age
  end
end
```

JunichiIto commented on 07.ls_object/file_detail.rb:
```diff
@@ -0,0 +1,58 @@
+    result = []
+    # ファイルモードを8進数に変換して末尾3桁（パーミッション）を取得する
+    stat.mode.to_s(8)[-3..].each_char do |char|
+      result << MODE_LIST[char.to_i]
```
mapに置き換えられそうです。

JunichiIto commented on 07.ls_object/file_detail.rb:
```diff
@@ -0,0 +1,58 @@
+
+  private
+
+  def fetch_file_mode(stat)
```
fetchは「プログラムの外部（OSのファイルシステムやDB、web APIなど）からデータを取ってくる」というニュアンスで使われることが多いので、この場合はfetch以外の名前が望ましい気がします〜。

rira100000000 replied:
`get_file_mode`に修正します。

JunichiIto reviewed (commented):
コメントしました〜。

rira100000000 committed:
AllCopsの削除

必要なくなったため、AllCopsを削除しました。

rira100000000 committed:
メソッド名、モジュール名の修正

メソッド名、モジュール名を適切なものに修正しました。

rira100000000 committed:
グローバルで定数を宣言しないよう修正

グローバルで定数を宣言していたため、クラス内で宣言するようにしました。

rira100000000 committed:
クラス設計の修正

LongFormatListとShortFormatListクラスはフォーマットだけを行うように修正しました。
ファイル情報の取得はFileDetailFetcherクラスで行うようにしました。

rira100000000 committed:
rオプションが正常に表示されない不具合修正

rira100000000 committed:
transposeを使うように修正

ファイル一覧を作成する際にtransposeメソッドを使うよう修正しました。

rira100000000 committed:
非ascii文字のずれを修正

非ascii文字を含むと列が乱れる不具合を修正しました。

rira100000000 committed:
テスト用ディレクトリの作成、

テスト用ディレクトリを作成し、.gitignoreに追加しました。

rira100000000 committed:
mapを使うように修正

each内で配列を作っていたのでmapを使うように修正しました。

rira100000000 committed:
代入方法を統一

一つのif文で別々の代入方法を使っていたので統一しました。

rira100000000 committed:
ループの引数名を修正

配列namesからfileを取り出していたので、nameを取り出すようにしました。

rira100000000 committed:
不要なインスタンス変数を削除

不要なインスタンス変数を削除して、呼び出し用にメソッドを追加しました。

rira100000000 committed:
不要なインスタンス変数を削除

不要なインスタンス変数を削除しました。

rira100000000 committed:
変数呼び出し用にメソッドを追加

uidとgidを呼び出すためのメソッドを追加しました。

rira100000000 committed:
FileDetailfetcherをls.rbでrequireする

long_format_listとshort_dormat_listでそれぞれrequireしていたので修正しました。

rira100000000 committed:
変数名を修正

rira100000000 committed:
mapを使うように修正

rira100000000 committed:
max_length_hashをStructに変更

ハッシュで定義していたmax_length_hashをStructに変更しました。

JunichiIto commented on 07.ls_object/file_detail.rb:
```diff
@@ -34,24 +34,31 @@ def initialize(name, directory = '')
+  end
+
+  def mode
+    get_file_mode(@stat)
```
他のメソッドを呼ぶだけかつ、呼ばれた先のメソッドが他で呼ばれることはないなら、そのメソッドの実装をここにそのまま書いてしまった方が良いです。

JunichiIto commented on 07.ls_object/file_detail.rb:
```diff
@@ -34,24 +34,31 @@ def initialize(name, directory = '')
 
-  def fetch_file_mode(stat)
-    result = []
+  def get_file_mode(stat)
```
statはインスタンス変数なので引数で渡すメリットはあまりなさそうです。

JunichiIto commented on 07.ls_object/file_detail.rb:
```diff
@@ -34,24 +34,31 @@ def initialize(name, directory = '')
+    result = stat.mode.to_s(8)[-3..].each_char.map do |char|
+      MODE_LIST[char.to_i]
     end
     result.join
```
```suggestion
    stat.mode.to_s(8)[-3..].each_char.map do |char|
      MODE_LIST[char.to_i]
    end.join
```

こういう書き方もありといえばありです。（好みは分かれるかも）

JunichiIto commented on 07.ls_object/file_detail_fetcher.rb:
```diff
@@ -1,7 +1,7 @@
 
-module FetchFileDetail
-  def make_formatted_list(paths, options)
+class FileDetailfetcher
```
動詞+erのクラスは実質「モジュール」ですし、ls.rbの中でしか使われないので、ls.rbの中にただのメソッドとして定義すれば良さそう、と思いました。

JunichiIto commented on 07.ls_object/file_list.rb:
```diff
@@ -0,0 +1,33 @@
+# frozen_string_literal: true
+
+class FileList
+  attr_reader :name_list, :directory
```
どちらも使われていない属性のように見えます。

JunichiIto commented on 07.ls_object/file_list.rb:
```diff
@@ -0,0 +1,33 @@
+      names = Dir.glob('*', base: @path, flags: flag).sort
+    else
+      directory = ''
+      names = [@path]
```
pathがディレクトリじゃなかった場合のテストも必要になりそうです。（そこまでカバーすべきか、ということも要検討かも）

JunichiIto commented on 07.ls_object/file_list.rb:
```diff
@@ -0,0 +1,33 @@
+# frozen_string_literal: true
+
+class FileList
```
このクラスも結局`list_detail`でFileDetailの配列を返すだけの「モジュール」なので、クラス化する必要はないように思いました。

JunichiIto commented on 07.ls_object/file_list.rb:
```diff
@@ -0,0 +1,33 @@
+  end
+
+  def list_detail
+    @names.map do |name|
```
```suggestion
  def list_detail
    names, directory = parse_path
    names.map do |name|
```

って書いても処理的にはほとんど変わらないはずです。

処理的には変わらない、ということは`@names`や`@directory`はインスタンス変数化する意味はない、と言えます。

あと、`@names`や`@directory`がインスタンス変数になっていると、以下のような挙動になります。
この挙動が妥当かどうか、という観点でも`@names`や`@directory`をインスタンス変数にすべきかどうかの判断材料になります。

```ruby
file_list = FileList.new(...)
file_list.list_detail

# list_detailに含まれていたファイルをターミナル上で削除する

# 削除したファイルがまだ表示される
file_list.list_detail
```

JunichiIto commented on 07.ls_object/short_format_list.rb:
```diff
@@ -1,35 +1,35 @@
-    file_list = FileList.new(path, reverse_required, hidden_file_required)
-    adjust_list_for_display(file_list.name_list)
+  def format(file_details)
+    adjust_list_for_display(file_details)
```
他のメソッドを呼ぶだけのメソッドはメソッドの意味がないので、実装コードをそのままここに書いてしまいましょう。

JunichiIto commented on 07.ls_object/short_format_list.rb:
```diff
@@ -0,0 +1,56 @@
+    max_file_names = []
+    file_details.each_with_index do |file_detail, i|
+      file_names << file_detail.name
+      next unless ((i + 1) % rows).zero? || i + 1 == file_details.length
```
この行、何をやってるんだろう？と3秒ほど考え込んでしまったので、可読性を改善した方が良さそうです〜。

https://twitter.com/jnchito/status/1326468865103073280

JunichiIto commented on 07.ls_object/short_format_list.rb:
```diff
@@ -0,0 +1,56 @@
+      file_names.fill('', file_names.length...rows)
+      file_names_list << file_names
+      file_names = []
+    end
```
21〜30行目のロジックはもしかすると each_slice を使うと簡単に書けるかも？
（他にもArrayクラスやEnumerableモジュールの公式ドキュメントを上から下までじっくり眺めると、いいメソッドが他にも見つかるかも？）

JunichiIto commented on 07.ls_object/long_format_list.rb:
```diff
@@ -0,0 +1,54 @@
+  attr_accessor :max_length_hash
+
+  def initialize
+    max_length = Struct.new(:nlink, :uid, :gid, :file_size, :file_name)
```
Ruby 3.2を使ってるならDataクラスの方が新しくて良いかもしれません。

JunichiIto commented on 07.ls_object/long_format_list.rb:
```diff
@@ -0,0 +1,54 @@
+
+  def initialize
+    max_length = Struct.new(:nlink, :uid, :gid, :file_size, :file_name)
+    @max_length = max_length.new(0, 0, 0, 0, 0)
```
```ruby
formatter = LongFormatList.new
formatter.format(file_details)
formatter.format(other_file_details)
```

みたいな呼び出し方をされると最大長の判定を間違えそうです。

インスタンス変数は`user.name`や`user.date_of_birth`のように、newされてから変化しない値になるのが原則です。

https://qiita.com/jnchito/items/4d62693525f5023013cc#%E3%81%98%E3%82%83%E3%81%82%E3%82%A4%E3%83%B3%E3%82%B9%E3%82%BF%E3%83%B3%E3%82%B9%E5%A4%89%E6%95%B0%E3%81%AF%E3%81%84%E3%81%A4%E4%BD%BF%E3%81%88%E3%81%B0%E3%81%84%E3%81%84%E3%81%AE

JunichiIto commented on 07.ls_object/long_format_list.rb:
```diff
@@ -0,0 +1,54 @@
+  end
+
+  def update_max_length(file_detail)
+    @max_length.nlink = [file_detail.stat.nlink.to_s.length, @max_length.nlink].max
```
デメテルの法則に反しない形で実装するのが良いと思います。

```suggestion
    @max_length.nlink = [file_detail.nlink.to_s.length, @max_length.nlink].max
```

JunichiIto replied:
デメテルの法則に反しない形で実装するのが良いと思います。

```suggestion
    @max_length.nlink = [file_detail.nlink.to_s.length, @max_length.nlink].max
```

JunichiIto commented on 07.ls_object/long_format_list.rb:
```diff
@@ -0,0 +1,54 @@
+require_relative 'file_list'
+
+class LongFormatList
+  attr_accessor :max_length_hash
```
使われてなさそうです。

JunichiIto commented on 07.ls_object/long_format_list.rb:
```diff
@@ -0,0 +1,54 @@
+# frozen_string_literal: true
+
+require_relative 'file_detail'
+require_relative 'file_list'
```
ここでrequireする必要はなさそうです。
全体的に見直してみてください〜。

https://bootcamp.fjord.jp/questions/1640

JunichiIto commented on 07.ls_object/long_format_list.rb:
```diff
@@ -0,0 +1,54 @@
+
+    result = []
+    result << "total #{calc_total_block(file_details)}"
+    result << adjust_list_for_display(file_details)
```
最終的にはjoinするので結果オーライにはなるのですが、`[1, [2, 3]]`のようにネストした配列ができてしまうのがちょっと気持ち悪いなと思いました。

JunichiIto commented on 07.ls_object/file_detail.rb:
```diff
@@ -0,0 +1,65 @@
+
+  def initialize(name, directory = '')
+    @name = name
+    path = if directory == ''
```
```suggestion
    path = if directory.empty?
```

みたいに書くとRubyらしいかも（好みの問題）

JunichiIto commented on 07.ls_object/file_detail.rb:
```diff
@@ -0,0 +1,65 @@
+    path = if directory == ''
+             name
+           else
+             "#{directory}/#{name}"
```
パスの連結はPath.joinメソッドを使うと少し安全に連結できます。

JunichiIto reviewed (commented):
コメントしました！

rira100000000 committed:
modeメソッドにget_file_modeの処理を移譲

rira100000000 committed:
FileDetailfetcherクラスの削除

FileDetailfetcherクラスを削除し、メソッドをls.rbに移しました。

rira100000000 committed:
不要なrequireと属性を削除しました

rira100000000 committed:
デメテルの法則に則るように修正

statから呼び出している属性をfile_detailクラスのメソッドとして呼び出せるようにしました。

rira100000000 committed:
file_list,rbを削除

file_list.rbを削除し、必要な処理はls.rbに移動しました

rira100000000 committed:
変数名の変更とコメントの追加

可読性を高めるため、変数名の変更とコメントを追加しました。

rira100000000 committed:
each_sliceを使ってリファクタリング

each_sliceを使って処理を書き直しました。

rira100000000 committed:
join前の配列を綺麗にする

join前の配列が[1,[2,3]]のような形になっていたので綺麗な1次配列になるようにしました。

rira100000000 committed:
.ruby-versionを追加

rira100000000 committed:
LongFormatListを`new`するタイミングでfile_detailsを渡すように修正

LongFormatListを`new`するタイミングでfile_detailsを渡すようにしました。
同様にShortFormatListでも`new`するタイミングでfile_detailsを渡すようにしました。

rira100000000 committed:
パスの連結にPathname.joinを使うように修正しました。

rira100000000 committed:
空文字列の判定にempty?を使うよう修正

空文字列の判定に=''を使用していたので、empty?に修正しました。

rira100000000 committed:
パスの指定機能を削除

必須要件出なかったのでパスの指定機能は削除しました

rira100000000 committed:
file_detailのinitializeにおける初期値の削除

initializeでdirectoryは必ず指定されるため、初期値を削除しました。

rira100000000 committed:
コマンドライン引数のパスを無視するよう修正

コマンドライン引数はa,l,rのみサポートするように修正しました。

rira100000000 committed:
formatメソッドをprivateに修正

外部から呼び出さないのでprivateに修正しました。

rira100000000 committed:
不要な変数の削除

calc_max_lengthをsplat展開し変数を作らないよう修正しました。

rira100000000 commented:
こちらに移しました。
[lsコマンドオブジェクト指向版　修正 by rira100000000 · Pull Request \#13 · rira100000000/ruby\-practices](https://github.com/rira100000000/ruby-practices/pull/13)
